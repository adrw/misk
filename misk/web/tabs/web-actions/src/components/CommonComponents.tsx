import { Collapse, Icon, Menu, MenuItem, Tag, Tooltip } from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import { HTTPMethodIntent } from "@misk/core"
import {
  ISimpleFormState,
  onChangeToggleFnCall,
  simpleSelect
} from "@misk/simpleredux"
import HTTPMethod from "http-method-enum"
import * as React from "react"
import { connect } from "react-redux"
import styled from "styled-components"
import { IDispatchProps, mapDispatchToProps, mapStateToProps } from "../ducks"

export const FloatLeft = styled.span`
  float: left;
  margin: 0 10px 0 0;
`

export const FloatRight = styled.span`
  float: right;
  margin: 0 0 0 10px;
`

export const CodeTag = styled(Tag)`
  font-family: monospace;
`

export const Header = styled.div`
  display: inline-block;
`

export const Column = styled.div`
  flex-grow: 1;
  flex-basis: 0;
  min-width: 320px;
`

/**
 * Renders HTTP Method tags for each Web Action card
 */
export const MethodTag = (props: { method: HTTPMethod }) => (
  <FloatRight>
    <Tag large={true} intent={HTTPMethodIntent[props.method]}>
      {props.method}
    </Tag>
  </FloatRight>
)

export const MetadataMenu = styled(Menu)`
  li {
    margin-bottom: 0;
  }
`

/**
 * Single cell component for a piece of metadata
 * @param props
 *  * content: primary content that is displayed (ie. eng, service-owners)
 *  * label: right aligned faded text that is a label for the data (ie. Roles)
 *  * tooltip: text displayed in tooltip when content is hovered over.
 *      Separate definition from content in the case that a truncated length
 *      is displayed but the full length of content is displayed in tooltip
 */
export const Metadata = (props: {
  content: string | JSX.Element
  label?: string
  labelElement?: JSX.Element
  onClick?: (event: any) => void
  tooltip?: string | JSX.Element
}) => {
  if (props.tooltip) {
    return (
      <MenuItem
        {...props}
        text={<Tooltip content={props.tooltip}>{props.content}</Tooltip>}
      />
    )
  } else {
    return <MenuItem {...props} text={<Tooltip>{props.content}</Tooltip>} />
  }
}

/**
 * Metadata that slides out content below when clicked
 * @param props : includes same props as Metadata with a few additional
 *  * children: any components to display when the Metadata is clicked
 *  * tag: string to use in @misk/SimpleRedux/SimpleForm to register Metadata clicks
 *  * IState: include connected State from parent container
 *      Provides access to @misk/SimpleRedux/SimpleForm substate in Redux
 *  * IDispatchProps: include connected dispatch object from parent container
 *      Provides access to @misk/SimpleRedux/SimpleForm input handlers
 */
const UnconnectedMetadataCollapse = (
  props: {
    children: any
    content: string | JSX.Element
    isOpen?: boolean
    label?: string
    labelElement?: JSX.Element
    simpleForm: ISimpleFormState
    tag: string
  } & IDispatchProps
) => {
  const { children } = props
  return (
    <div>
      <Metadata
        content={
          <span>
            {simpleSelect(props.simpleForm, props.tag, "data") ? (
              <Icon icon={IconNames.CARET_DOWN} />
            ) : (
              <Icon icon={IconNames.CARET_RIGHT} />
            )}{" "}
            {props.content}
          </span>
        }
        data-testid={"metadata-collapse"}
        label={props.label}
        labelElement={props.labelElement}
        onClick={onChangeToggleFnCall(
          props.simpleFormToggle,
          props.tag,
          props.simpleForm
        )}
      />
      <Collapse isOpen={simpleSelect(props.simpleForm, props.tag, "data")}>
        {children}
      </Collapse>
    </div>
  )
}

export const MetadataCollapse = connect(
  mapStateToProps,
  mapDispatchToProps
)(UnconnectedMetadataCollapse)
