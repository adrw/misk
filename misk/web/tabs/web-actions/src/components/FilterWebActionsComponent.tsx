import { ControlGroup, HTMLSelect, InputGroup } from "@blueprintjs/core"
import { onChangeFnCall } from "@misk/simpleredux"
import * as React from "react"
import { IDispatchProps, IState, WebActionInternalLabel } from "../ducks"

export const FilterWebActionsComponent = (
  props: IState & IDispatchProps & { tag: string }
) => {
  const FilterSelectOptions = [...Object.keys(WebActionInternalLabel)]
  const filterTag = `${props.tag}::Filter`
  return (
    <div>
      <ControlGroup fill={true}>
        <HTMLSelect
          large={true}
          onChange={onChangeFnCall(
            props.simpleFormInput,
            `${filterTag}::HTMLSelect`
          )}
          options={FilterSelectOptions}
        />
        <InputGroup
          large={true}
          onChange={onChangeFnCall(
            props.simpleFormInput,
            `${filterTag}::Input`
          )}
          placeholder={"Filter Web Actions"}
        />
        {/* TODO(adrw) Use multiselect to do autocomplete filters */}
      </ControlGroup>
    </div>
  )
}
