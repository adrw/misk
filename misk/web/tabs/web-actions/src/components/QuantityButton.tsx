import { Button, Tooltip } from "@blueprintjs/core"
import { IconName } from "@blueprintjs/icons"
import { onChangeFnCall } from "@misk/simpleredux"
import * as React from "react"
import { IWebActionInternal } from "../ducks"

export const QuantityButton = (props: {
  id: string
  action: IWebActionInternal
  changeFn: any
  content: string
  icon: IconName
  oldState: any
}) => (
  <Tooltip content={props.content}>
    <Button
      icon={props.icon}
      onClick={onChangeFnCall(
        props.changeFn,
        props.id,
        props.action,
        props.oldState
      )}
    />
  </Tooltip>
)
