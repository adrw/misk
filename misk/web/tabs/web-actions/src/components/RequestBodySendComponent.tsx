import {
  Button,
  Collapse,
  ControlGroup,
  HTMLSelect,
  Label,
  Tag
} from "@blueprintjs/core"
import {
  CodePreContainer,
  HTTPMethodDispatch,
  HTTPMethodIntent
} from "@misk/core"
import {
  onChangeFnCall,
  onClickFnCall,
  simpleSelect,
  simpleType
} from "@misk/simpleredux"
import * as React from "react"
import {
  IDispatchProps,
  IFieldTypeMetadata,
  IState,
  IWebActionInternal
} from "../ducks"
import {
  generateKeyTag,
  parseWebActionConstants,
  safeNumberArray
} from "./utilities"

const selectTypedData = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { requestType, types } = props.action
  const { tag } = props
  if (requestType && types && types[requestType] && types[requestType].fields) {
    const { fields } = types[requestType]
    const fieldsData = fields.map((field: IFieldTypeMetadata) => {
      const fieldKeys = safeNumberArray(
        simpleSelect(
          props.simpleForm,
          generateKeyTag({ field, nestPath: "/", tag }),
          "data",
          simpleType.array
        )
      )
      return fieldKeys.map((key: number) => {
        const keyData = simpleSelect(
          props.simpleForm,
          `${tag}::/${key}::Data`,
          "data"
        )
        console.log(field.name, key, keyData)
        return keyData
      })
    })
    return fieldsData
  } else {
    return []
  }
}

export const RequestBodySendComponent = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { requestType, types } = props.action
  const { url, method, tag } = parseWebActionConstants(props)
  const typedData = selectTypedData(props)
  if (requestType && types && types[requestType] && types[requestType].fields) {
    return (
      <div>
        <ControlGroup>
          <HTMLSelect
            large={true}
            onChange={onChangeFnCall(props.simpleFormInput, `${tag}::Method`)}
            options={props.action.dispatchMechanism.sort()}
            value={method}
          />
          <Button
            large={true}
            onClick={onClickFnCall(
              HTTPMethodDispatch(props)[method],
              `${tag}::Response`,
              url,
              typedData
            )}
            intent={HTTPMethodIntent[method]}
            loading={simpleSelect(
              props.simpleNetwork,
              `${tag}::Response`,
              "loading"
            )}
            text={"Submit"}
          />
        </ControlGroup>
        <Label>
          Request <Tag>{url}</Tag>
        </Label>
        <Collapse isOpen={typedData.length > 0}>
          <CodePreContainer>
            {JSON.stringify(typedData, null, 2)}
          </CodePreContainer>
        </Collapse>
      </div>
    )
  } else {
    return (
      <div>
        <ControlGroup>
          <HTMLSelect
            large={true}
            onChange={onChangeFnCall(props.simpleFormInput, `${tag}::Method`)}
            options={props.action.dispatchMechanism.sort()}
            value={method}
          />
          <Button
            large={true}
            onClick={onClickFnCall(
              HTTPMethodDispatch(props)[method],
              `${tag}::Response`,
              url,
              simpleSelect(props.simpleForm, `${tag}::Body`, "data")
            )}
            intent={HTTPMethodIntent[method]}
            loading={simpleSelect(
              props.simpleNetwork,
              `${tag}::Response`,
              "loading"
            )}
            text={"Submit"}
          />
        </ControlGroup>
        <Label>
          Request <Tag>{url}</Tag>
        </Label>
        <Collapse
          isOpen={simpleSelect(props.simpleForm, `${tag}::Body`, "data")}
        >
          <CodePreContainer>
            {JSON.stringify(
              simpleSelect(props.simpleForm, `${tag}::Body`, "data"),
              null,
              2
            )}
          </CodePreContainer>
        </Collapse>
      </div>
    )
  }
}
