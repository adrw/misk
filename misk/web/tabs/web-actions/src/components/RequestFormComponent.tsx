import {
  Button,
  Card,
  ControlGroup,
  FormGroup,
  InputGroup,
  Intent,
  TextArea,
  Tooltip
} from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import {
  onChangeFnCall,
  onChangeToggleFnCall,
  simpleSelect,
  simpleType
} from "@misk/simpleredux"
import { Map } from "immutable"
import * as React from "react"
import styled from "styled-components"
import {
  BaseFieldTypes,
  findIndexAction,
  IDispatchProps,
  IState,
  ITypesFieldMetadata,
  IWebActionInternal,
  padId,
  TypescriptBaseTypes
} from "../ducks"

const RequestFieldGroup = styled(Card)`
  box-shadow: none !important;
  border-left: 2px black;
  margin-bottom: 0px;
  padding: 0px 0px 0px 10px !important;
`

const RequestFormGroup = styled(FormGroup)`
  margin: 0 !important;
`

const AddButton = (
  props: {
    idParent: string
    action: IWebActionInternal
  } & IState &
    IDispatchProps
) => (
  <Tooltip content={"Add Field"}>
    <Button
      icon={IconNames.PLUS}
      onClick={onChangeFnCall(
        props.webActionsAdd,
        props.idParent,
        props.action,
        props.webActionsRaw
      )}
    />
  </Tooltip>
)

const RemoveButton = (
  props: {
    id: string
    action: IWebActionInternal
  } & IState &
    IDispatchProps
) => (
  <Tooltip content={"Remove Field"}>
    <Button
      icon={IconNames.CROSS}
      onClick={onChangeFnCall(
        props.webActionsRemove,
        props.id,
        props.action,
        props.webActionsRaw
      )}
    />
  </Tooltip>
)

const repeatableFieldButtons = (
  props: {
    action: IWebActionInternal
    id: string
    typesMetadata: Map<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => {
  const { action, id, typesMetadata } = props
  const metadata = typesMetadata.get(id)
  if (
    metadata &&
    metadata.id !== "0" &&
    typesMetadata.get(metadata.idParent).repeated
  ) {
    const { idParent, name, serverType } = metadata
    const { idChildren } = typesMetadata.get(idParent)
    const removeButton =
      idChildren.size > 1 ? (
        <RemoveButton {...props} action={action} id={id} />
      ) : (
        <span />
      )
    return [
      <Tooltip content={`Repeated ${serverType}`}>
        <Button icon={IconNames.REPEAT}>{name}</Button>
      </Tooltip>,
      <AddButton {...props} action={action} idParent={idParent} />,
      removeButton
    ]
  } else if (metadata && !metadata.repeated) {
    const { name, serverType } = metadata
    return [
      <Tooltip content={serverType}>
        <Button>{name || "Body"}</Button>
      </Tooltip>
    ]
  } else {
    return [<span />]
  }
}

const RequestFormFieldBuilder = (
  props: {
    action: IWebActionInternal
    id: string
    tag: string
    typesMetadata: Map<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => {
  const { id, tag, typesMetadata } = props
  const metadata = typesMetadata.get(id)
  const {
    idChildren,
    name,
    serverType,
    typescriptType
  } = metadata as ITypesFieldMetadata
  if (typescriptType === null) {
    if (
      idChildren.first() &&
      typesMetadata.get(idChildren.first()).typescriptType &&
      BaseFieldTypes.hasOwnProperty(serverType)
    ) {
      return (
        <div>
          {idChildren.map((child: string) => (
            <RequestFormFieldBuilder {...props} id={child} />
          ))}
        </div>
      )
    } else {
      const fieldGroup = (child: string) => {
        const { serverType: childServerType } = typesMetadata.get(child)
        if (BaseFieldTypes.hasOwnProperty(childServerType)) {
          return (
            <div>
              <RequestFormFieldBuilder {...props} id={child} />
            </div>
          )
        } else {
          return (
            <div>
              <ControlGroup>
                {...repeatableFieldButtons({ ...props, id: child })}
              </ControlGroup>
              <RequestFormFieldBuilder {...props} id={child} />
            </div>
          )
        }
      }
      return (
        <RequestFieldGroup>
          {idChildren.map((child: string) => fieldGroup(child))}
        </RequestFieldGroup>
      )
    }
  } else if (typescriptType === TypescriptBaseTypes.boolean) {
    return (
      <ControlGroup>
        {...repeatableFieldButtons({ ...props, id })}
        <Button
          intent={
            simpleSelect(
              props.simpleForm,
              `${tag}::${padId(id)}`,
              "data",
              simpleType.boolean
            )
              ? Intent.PRIMARY
              : Intent.WARNING
          }
          onClick={onChangeToggleFnCall(
            props.simpleFormToggle,
            `${tag}::${padId(id)}`,
            props.simpleForm
          )}
        >
          {simpleSelect(
            props.simpleForm,
            `${tag}::${padId(id)}`,
            "data",
            simpleType.boolean
          ).toString()}
        </Button>
      </ControlGroup>
    )
  } else if (typescriptType === TypescriptBaseTypes.number) {
    return (
      <ControlGroup>
        {...repeatableFieldButtons({ ...props, id })}
        <InputGroup
          onChange={onChangeFnCall(
            props.simpleFormInput,
            `${tag}::${padId(id)}`
          )}
          placeholder={serverType}
          value={simpleSelect(props.simpleForm, `${tag}::${padId(id)}`, "data")}
        />
      </ControlGroup>
    )
  } else if (typescriptType === TypescriptBaseTypes.string) {
    return (
      <ControlGroup>
        {...repeatableFieldButtons({ ...props, id })}
        <Tooltip content={"Toggle large text input"}>
          <Button
            active={simpleSelect(
              props.simpleForm,
              `${tag}::LongText:${padId(id)}`,
              "data",
              simpleType.boolean
            )}
            icon={IconNames.MORE}
            onClick={onChangeToggleFnCall(
              props.simpleFormToggle,
              `${tag}::LongText:${padId(id)}`,
              props.simpleForm
            )}
          />
        </Tooltip>
        {simpleSelect(
          props.simpleForm,
          `${tag}::LongText:${padId(id)}`,
          "data",
          simpleType.boolean
        ) ? (
          <TextArea
            fill={true}
            onChange={onChangeFnCall(
              props.simpleFormInput,
              `${tag}::${padId(id)}`
            )}
            placeholder={`${serverType}\nDrag bottom right corner of text area input to expand.`}
            value={simpleSelect(
              props.simpleForm,
              `${tag}::${padId(id)}`,
              "data"
            )}
          />
        ) : (
          <InputGroup
            onChange={onChangeFnCall(
              props.simpleFormInput,
              `${tag}::${padId(id)}`
            )}
            placeholder={serverType}
            value={simpleSelect(
              props.simpleForm,
              `${tag}::${padId(id)}`,
              "data"
            )}
          />
        )}
      </ControlGroup>
    )
  } else {
    return (
      <div>
        {...repeatableFieldButtons({ ...props, id })}
        <RequestFormGroup label={`${name}:${JSON.stringify(serverType)}`}>
          <TextArea
            fill={true}
            onChange={onChangeFnCall(
              props.simpleFormInput,
              `${tag}::${padId(id)}`
            )}
            placeholder={
              "Unparseable type. (JSON or Text).\nDrag bottom right corner of text area input to expand."
            }
          />
        </RequestFormGroup>
      </div>
    )
  }
}

export const RequestFormComponent = (
  props: {
    action: IWebActionInternal
    tag: string
  } & IState &
    IDispatchProps
) => {
  const { action } = props
  const { typesMetadata } = props.webActionsRaw.get("metadata")[
    findIndexAction(action, props.webActionsRaw)
  ]
  return (
    <RequestFormFieldBuilder
      {...props}
      id={"0"}
      typesMetadata={typesMetadata}
    />
  )
}
