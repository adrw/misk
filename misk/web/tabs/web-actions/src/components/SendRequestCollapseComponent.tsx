import {
  Button,
  Collapse,
  ControlGroup,
  HTMLSelect,
  InputGroup
} from "@blueprintjs/core"
import {
  CodePreContainer,
  FlexContainer,
  HTTPMethodDispatch,
  HTTPMethodIntent
} from "@misk/core"
import { onChangeFnCall, simpleSelect } from "@misk/simpleredux"
import { HTTPMethod } from "http-method-enum"
import * as React from "react"
import styled from "styled-components"
import {
  Metadata,
  MetadataCollapse,
  MetadataMenu,
  RequestFormComponent,
  StatusTagComponent
} from "../components"
import {
  getFormData,
  IDispatchProps,
  IState,
  IWebActionInternal,
  methodHasBody
} from "../ducks"

const Column = styled.div`
  flex-grow: 1;
  flex-basis: 0;
  min-width: 320px;
`

const RequestBodyForm = (
  props: {
    action: IWebActionInternal
    method: HTTPMethod
    tag: string
  } & IState &
    IDispatchProps
) => {
  const { method, tag } = props
  if (methodHasBody(props.method)) {
    return (
      <MetadataCollapse
        content={"Request Body"}
        label={"Input"}
        tag={`${tag}::ButtonFormRequestBody`}
      >
        <RequestFormComponent {...props} tag={tag} />
      </MetadataCollapse>
    )
  } else {
    return (
      <Metadata
        {...props}
        content={"Request Body"}
        label={`${method} does not have a Body`}
      />
    )
  }
}

/**
 * Collapse wrapped Send a Request form for each Web Action card
 */
export const SendRequestCollapseComponent = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { action, tag } = props
  // Determine if Send Request form for the Web Action should be open
  const isOpen = simpleSelect(
    props.simpleForm,
    `${tag}::ButtonSendRequest`,
    "data"
  )
  const url = simpleSelect(props.simpleForm, `${tag}::URL`, "data")
  // Pre-populate the URL field with the action path pattern on open of request form
  if (isOpen && !url) {
    props.simpleFormInput(`${tag}::URL`, action.pathPattern)
  }
  const method: HTTPMethod =
    simpleSelect(props.simpleForm, `${tag}::Method`, "data") ||
    action.dispatchMechanism.reverse()[0]
  const formData = getFormData(
    action,
    props.simpleForm,
    tag,
    props.webActionsRaw
  )
  if (
    methodHasBody(method) &&
    typeof simpleSelect(
      props.simpleForm,
      `${tag}::ButtonRequestBody`,
      "data"
    ) === "string"
  ) {
    props.simpleFormInput(`${tag}::ButtonRequestBody`, true)
  }
  if (
    methodHasBody(method) &&
    typeof simpleSelect(
      props.simpleForm,
      `${tag}::ButtonFormRequestBody`,
      "data"
    ) === "string"
  ) {
    props.simpleFormInput(`${tag}::ButtonFormRequestBody`, true)
  }
  if (
    typeof simpleSelect(props.simpleForm, `${tag}::ButtonResponse`, "data") ===
      "string" &&
    simpleSelect(props.simpleNetwork, `${tag}::Response`, "data")
  ) {
    props.simpleFormInput(`${tag}::ButtonResponse`, true)
  }
  return (
    <Collapse isOpen={isOpen}>
      <InputGroup
        defaultValue={action.pathPattern}
        onChange={onChangeFnCall(props.simpleFormInput, `${tag}::URL`)}
        placeholder={
          "Request URL: absolute ( http://your.url.com/to/send/a/request/to/ ) or internal service endpoint ( /service/web/action )"
        }
        type={"url"}
      />
      <FlexContainer>
        <Column>
          <MetadataMenu>
            <RequestBodyForm
              action={action}
              method={method}
              tag={tag}
              {...props}
            />
          </MetadataMenu>
        </Column>
        <Column>
          <ControlGroup>
            <HTMLSelect
              large={true}
              onChange={onChangeFnCall(props.simpleFormInput, `${tag}::Method`)}
              options={action.dispatchMechanism.sort()}
              value={method}
            />
            <Button
              large={true}
              onClick={(event: any) => {
                props.simpleFormInput(`${tag}::ButtonRequestBody`, false)
                HTTPMethodDispatch(props)[method](
                  `${tag}::Response`,
                  url,
                  isOpen && formData
                )
              }}
              intent={HTTPMethodIntent[method]}
              loading={simpleSelect(
                props.simpleNetwork,
                `${tag}::Response`,
                "loading"
              )}
              text={"Submit"}
            />
          </ControlGroup>
          <MetadataMenu>
            {methodHasBody(method) ? (
              <MetadataCollapse
                {...props}
                content={"Request Body"}
                label={`${url}`}
                tag={`${tag}::ButtonRequestBody`}
              >
                <CodePreContainer>
                  {JSON.stringify(isOpen && formData, null, 2)}
                </CodePreContainer>
              </MetadataCollapse>
            ) : (
              <Metadata content={"Request"} label={`${url}`} {...props} />
            )}
            <MetadataCollapse
              {...props}
              content={"Response"}
              labelElement={
                <StatusTagComponent
                  status={simpleSelect(
                    props.simpleNetwork,
                    `${props.tag}::Response`,
                    "status"
                  )}
                />
              }
              tag={`${tag}::ButtonResponse`}
            >
              <div>
                <CodePreContainer>
                  {JSON.stringify(
                    simpleSelect(
                      props.simpleNetwork,
                      `${tag}::Response`,
                      "data"
                    ),
                    null,
                    2
                  )}
                </CodePreContainer>
                <MetadataCollapse
                  {...props}
                  content={"Raw Response"}
                  label={"Redux State"}
                  tag={`${tag}::ButtonRawResponse`}
                >
                  <CodePreContainer>
                    {JSON.stringify(
                      simpleSelect(props.simpleNetwork, `${tag}::Response`),
                      null,
                      2
                    )}
                  </CodePreContainer>
                </MetadataCollapse>
              </div>
            </MetadataCollapse>
          </MetadataMenu>
        </Column>
      </FlexContainer>
    </Collapse>
  )
}
