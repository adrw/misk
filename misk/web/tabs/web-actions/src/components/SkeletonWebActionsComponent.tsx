import { Card, Classes, H3 } from "@blueprintjs/core"
import { FlexContainer } from "@misk/core"
import { HTTPMethod } from "http-method-enum"
import * as React from "react"
import {
  CodeTag,
  Column,
  FloatLeft,
  Header,
  Metadata,
  MetadataMenu,
  MethodTag
} from "."

/**
 * Empty Web Action Card UI for use with BlueprintJS Skeleton class in loading UIs
 * https://blueprintjs.com/docs/#core/components/skeleton
 */
export const SkeletonText = () => (
  <span className={Classes.SKELETON}>{"Lorem ipsum"}</span>
)

export const SkeletonWebActionsComponent = () => (
  <Card>
    <Header>
      {[HTTPMethod.GET].map((m, index) => (
        <MethodTag key={index} method={m} />
      ))}
      <FloatLeft>
        <H3 className={Classes.SKELETON}>{"AnotherSimpleWebAction"}</H3>
      </FloatLeft>
      <FloatLeft>
        <CodeTag large={true}>{<SkeletonText />}</CodeTag>
      </FloatLeft>
    </Header>
    <FlexContainer>
      <Column>
        <MetadataMenu>
          <Metadata label={"Function"} content={<SkeletonText />} />
          <Metadata label={"Services"} content={<SkeletonText />} />
          <Metadata label={"Roles"} content={<SkeletonText />} />
          <Metadata label={"Access"} content={<SkeletonText />} />
        </MetadataMenu>
      </Column>
      <Column>
        <MetadataMenu>
          <Metadata label={"Content Types"} content={<SkeletonText />} />
          <Metadata
            label={"Application Interceptors"}
            content={<SkeletonText />}
          />
          <Metadata label={"Network Interceptors"} content={<SkeletonText />} />
          <Metadata label={"Send a Request"} content={<SkeletonText />} />
        </MetadataMenu>
      </Column>
    </FlexContainer>
  </Card>
)
