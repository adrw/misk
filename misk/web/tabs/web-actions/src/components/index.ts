export * from "./CommonComponents"
export * from "./RequestFormComponent"
export * from "./RequestResponseContentTypes"
export * from "./SendRequestCollapseComponent"
export * from "./SkeletonWebActionsComponent"
export * from "./StatusTagComponent"
