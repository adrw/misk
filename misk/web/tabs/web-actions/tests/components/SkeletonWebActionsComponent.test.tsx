import React from "react"
import { cleanup, render } from "react-testing-library"
import { SkeletonText, SkeletonWebActionsComponent } from "../../src/components"

// TODO upstream these to @misk/core or @misk/test
jest.mock("@misk/core", () => {
  const miskCore = require.requireActual("@misk/core")
  const mock = {
    ...miskCore,
    Column: jest.fn((props: { children: any }) => (
      <div className={"Column"}>{props.children}</div>
    )),
    FlexContainer: jest.fn((props: { children: any }) => (
      <div className={"FlexContainer"}>{props.children}</div>
    ))
  }
  return mock
})

describe("SkeletonWebActionsComponent", () => {
  afterEach(cleanup)
  it("SkeletonText glowing text block used as loading screen", () => {
    const { asFragment } = render(<SkeletonText />)
    expect(asFragment()).toMatchSnapshot()
  })
  it("SkeletonWebActionsComponent single Web Actions Card used as loading screen", () => {
    const { asFragment } = render(<SkeletonWebActionsComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
})
