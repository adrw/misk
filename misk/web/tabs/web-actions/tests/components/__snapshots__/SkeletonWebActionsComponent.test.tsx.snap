// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkeletonWebActionsComponent SkeletonText glowing text block used as loading screen 1`] = `
<DocumentFragment>
  <span
    class="bp3-skeleton"
  >
    Lorem ipsum
  </span>
</DocumentFragment>
`;

exports[`SkeletonWebActionsComponent SkeletonWebActionsComponent single Web Actions Card used as loading screen 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0"
  >
    <div
      class="sc-bxivhb jzkpli"
    >
      <span
        class="sc-bwzfXH jyNnlm"
      >
        <span
          class="bp3-tag bp3-intent-primary bp3-large"
        >
          <span
            class="bp3-text-overflow-ellipsis bp3-fill"
          >
            GET
          </span>
        </span>
      </span>
      <span
        class="sc-bdVaJa YEVeM"
      >
        <h3
          class="bp3-heading bp3-skeleton"
        >
          AnotherSimpleWebAction
        </h3>
      </span>
      <span
        class="sc-bdVaJa YEVeM"
      >
        <span
          class="bp3-tag bp3-large sc-htpNat hpvuBC"
        >
          <span
            class="bp3-text-overflow-ellipsis bp3-fill"
          >
            <span
              class="bp3-skeleton"
            >
              Lorem ipsum
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      class="FlexContainer-sc-uyd4r0 eEoNQY"
    >
      <div
        class="sc-ifAKCX iOsfuG"
      >
        <ul
          class="bp3-menu sc-EHOje gDRVEN"
        >
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Function"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Function
              </span>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Services"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Services
              </span>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Roles"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Roles
              </span>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Access"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Access
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="sc-ifAKCX iOsfuG"
      >
        <ul
          class="bp3-menu sc-EHOje gDRVEN"
        >
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Content Types"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Content Types
              </span>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Application Interceptors"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Application Interceptors
              </span>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Network Interceptors"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Network Interceptors
              </span>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
              content="[object Object]"
              label="Send a Request"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <span
                      class="bp3-skeleton"
                    >
                      Lorem ipsum
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="bp3-menu-item-label"
              >
                Send a Request
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
